components/claim/components/email_to_ldap.tsx

```diff
@@ -11,81 +10,95 @@ import {t} from 'utils/i18n.jsx';
 import LoginMfa from 'components/login/login_mfa.jsx';
 import LocalizedInput from 'components/localized_input/localized_input';
 
-export default class EmailToLDAP extends React.PureComponent {
-    static propTypes = {
-        email: PropTypes.string,
-        siteName: PropTypes.string,
-        ldapLoginFieldName: PropTypes.string,
-    };
+type Props = {
+    email?: any,
+    siteName?: any,
+    ldapLoginFieldName?: any,
+};
 
-    constructor(props) {
-        super(props);
 
-        this.state = {
-            passwordError: '',
-            ldapError: '',
-            ldapPasswordError: '',
-            serverError: '',
-            showMfa: false,
-        };
+type State = {
+    passwordError?: string,
+    ldapError?: string,
+    ldapPasswordError?: string,
+    serverError?: string,
+    showMfa?: boolean,
+};
+
+export default class EmailToLDAP extends React.PureComponent <Props, State>{
+    private emailPasswordInput: React.RefObject<HTMLInputElement>;
+    private ldapIdInput: React.RefObject<HTMLInputElement>;
+    private ldapPasswordInput: React.RefObject<HTMLInputElement>;
+
+    public constructor(props: Props) {
+        super(props);
 
         this.emailPasswordInput = React.createRef();
         this.ldapIdInput = React.createRef();
         this.ldapPasswordInput = React.createRef();
+        this.state = this.getDefaultState();
     }
 
-    preSubmit = (e) => {
-        e.preventDefault();
-
-        var state = {
+    getDefaultState() {
+        return {
             passwordError: '',
             ldapError: '',
             ldapPasswordError: '',
             serverError: '',
+            ldapId: '',
+            password: '',
+            ldapPassword: '',
```

Comment:
missing `showMfa`, I see some other fields not present in type state.

if we are initializing to the empty string, why did we add `?` to the types?

Reply:
I'm correcting this

Reply:
So I need to list all other fields like `showMfa` and also deleting `?` in type state like this?
![Screen Shot 2021-06-23 at 11 42 04](https://user-images.githubusercontent.com/35423212/123032155-b028e480-d40f-11eb-92f4-d8f441f8a2e0.png)

Reply:
I asked about '?' because I'm not sure it can be null, if it can't we shouldn't allow it, so the compiler knows about it.

Reply:
Yes I agree with you the type State cannot be null, but their defaultState is null